### PR TITLE
Adds CoM randomization term to manager-based events

### PR DESCRIPTION
**Pull request by shendredm**
# Description

Introduced CoM randomization in events.py for manager based envs. Individual range can be selected for individual axis and bodies.  
Thanks to #641 and #693  for motivation of the code.  

## Type of change

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

## Screenshots

Please attach before and after screenshots of the change if applicable.

## Checklist

- [x] I have run the [`pre-commit` checks](https://pre-commit.com/) with `./isaaclab.sh --format`
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have updated the changelog and the corresponding version in the extension's `config/extension.toml` file
- [x] I have added my name to the `CONTRIBUTORS.md` or my name already exists there

### CODE REVIEW

File: source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/events.py [hunk 1]
```diff
@@ -1004,3 +1004,46 @@ def _randomize_prop_by_op(
             f"Unknown operation: '{operation}' for property randomization. Please use 'add', 'scale', or 'abs'."
         )
     return data
+
+def randomize_body_coms(
```
Comment (Mayankm96) [inline source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/events.py]: Could you please place this after: randomize_rigid_body_mass function to keep ordering consistent to follow?

File: source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/events.py [hunk 2]
```diff
@@ -1004,3 +1004,46 @@ def _randomize_prop_by_op(
...
+    # resolve body indices
+    if asset_cfg.body_ids == slice(None):
+        body_ids = torch.arange(asset.num_bodies, dtype=torch.int, device="cpu")
+    else:
+        body_ids = torch.tensor(asset_cfg.body_ids, dtype=torch.int, device="cpu")
+
+    # sample random CoM values
+    range_list = [com_range.get(key, (0.0, 0.0)) for key in ["x", "y", "z"]]
+    ranges = torch.tensor(range_list, device="cpu")
+    rand_samples = math_utils.sample_uniform(ranges[:, 0], ranges[:, 1], (len(env_ids), 3), device="cpu").unsqueeze(1)
```
Comment (Mayankm96) [inline source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/events.py]: This should be num bodies too?

```suggestion
    rand_samples = math_utils.sample_uniform(ranges[:, 0], ranges[:, 1], (len(env_ids), len(body_ids), 3), device="cpu").unsqueeze(1)
```

File: source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/events.py [hunk 3]
```diff
@@ -1004,3 +1004,46 @@ def _randomize_prop_by_op(
...
+
+    # get the current com of the bodies (num_assets, num_bodies)
+    coms = asset.root_physx_view.get_coms().clone()[:, body_ids, :3]
+
+    # Randomize the com in range
+    coms += rand_samples
+
+    # Set the new coms
+    new_coms = asset.root_physx_view.get_coms().clone()
+    new_coms[:, asset_cfg.body_ids, 0:3] = coms
```
Comment (Mayankm96) [inline source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/events.py]:
```suggestion
    coms = asset.root_physx_view.get_coms().clone()

    # Randomize the com in range
    coms[:, body_ids, :3] += rand_samples

    # Set the new coms
```

File: source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/events.py [hunk 4]
```diff
@@ -1004,3 +1004,46 @@ def _randomize_prop_by_op(
...
+    asset.root_physx_view.set_coms(new_coms, env_ids)
```
Comment (Mayankm96) [inline source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/events.py]:
```suggestion
    asset.root_physx_view.set_coms(coms, env_ids)
```

File: source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/events.py [hunk 5]
```diff
@@ -1004,3 +1004,46 @@ def _randomize_prop_by_op(
...
         )
     return data
+
+def randomize_body_coms(
+    env: ManagerBasedEnv, 
+    env_ids: torch.Tensor | None, 
+    com_range: dict[str, tuple[float, float]], 
+    asset_cfg: SceneEntityCfg
+):
+    """Randomize the CoM of the bodies by adding a random value sampled from the given range.
```
Comment (Mayankm96) [inline source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/events.py]:
```suggestion
    """Randomize the center of mass (CoM) of rigid bodies by adding a random value sampled from the given ranges.
```

File: source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/events.py (hunk 1 above)
Comment (Mayankm96) [inline source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/events.py]:
```suggestion
def randomize_rigid_body_com(
```

File: source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/manager_based/locomotion/velocity/velocity_env_cfg.py [hunk 6]
```diff
@@ -172,6 +172,15 @@ class EventCfg:
             "operation": "add",
         },
     )
+    
+    com_randomize = EventTerm(
```
Comment (Mayankm96) [inline source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/manager_based/locomotion/velocity/velocity_env_cfg.py]:
```suggestion
    base_com = EventTerm(
```

File: source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/manager_based/locomotion/velocity/velocity_env_cfg.py [hunk 7]
```diff
@@ -172,6 +172,15 @@ class EventCfg:
...
+        func=mdp.randomize_body_coms,
+        mode="startup",
+        params={
+            "asset_cfg": SceneEntityCfg("robot", body_names="trunk"),
```
Comment (Mayankm96) [inline source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/manager_based/locomotion/velocity/velocity_env_cfg.py]: Not all robots have bodies named trunk. In the default configuration, it is called "base" and we set it to "trunk" wherever applicable (inside Unitree Go1, Go2, etc configs). Please do the same and also report the training plots to make sure it isn't breaking the expected convergence.

File: CONTRIBUTORS.md [hunk 8]
```diff
@@ -88,6 +88,7 @@ Guidelines for modifications:
 * Yujian Zhang
 * Zhengyu Zhang
 * Ziqi Fan
+* Dhananjay Shendre
```
Comment (Mayankm96) [inline CONTRIBUTORS.md]: Please follow the instructions and put your name in alphabetical order.

**Review by Mayankm96 — CHANGES_REQUESTED** [review]
Thanks a lot for making the MR. Please address the comments in the review

**Review by shendredm — COMMENTED** [review]
Hi , I've made the suggested changes. Could you please review them when you get a chance? Thanks

**Review by jtigue-bdai — APPROVED** [review]
LGTM, thanks!

**Comment by shendredm** [comment]
Hi, I have modified this MR for IsaacLab 2.0.0.  Could you please review them when you get a chance? Thanks

**Comment by ParlitsisG** [comment]
Hi, @shendredm,

I noticed that this method changes the center of mass (COM), but how does it adjust the inertia? Changing the COM from the default value seems to imply a non-uniform density distribution. I plan to use this method for a project involving the manipulation of a cube with non-uniform density, and I want to modify the inertial parameters accordingly. Could you please clarify if this approach remains physically consistent with reality?

File: source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/manager_based/locomotion/velocity/velocity_env_cfg.py (hunk 6 above)
Comment (kellyguo11) [inline source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/manager_based/locomotion/velocity/velocity_env_cfg.py]: Is it necessary to have this added to the environment here? Could we make it a separate unit test instead?

File: source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/events.py (hunk 4 above)
Comment (kellyguo11) [inline source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/events.py]: could you please also run the formatter?

File: source/isaaclab/isaaclab/envs/mdp/events.py [hunk 9]
```diff
@@ -236,6 +236,47 @@ def randomize_rigid_body_mass(
...
+    rand_samples = math_utils.sample_uniform(ranges[:, 0], ranges[:, 1], (len(env_ids), len(body_ids), 3), device="cpu").unsqueeze(1)
+
+    # get the current com of the bodies (num_assets, num_bodies)
+    coms = asset.root_physx_view.get_coms().clone()
+
+    # Randomize the com in range
+    coms[:, body_ids, :3] += rand_samples
+
+    # Set the new coms
+    asset.root_physx_view.set_coms(new_coms, env_ids)
```
Comment (kellyguo11) [inline source/isaaclab/isaaclab/envs/mdp/events.py]:
```suggestion
    asset.root_physx_view.set_coms(coms, env_ids)
```

File: source/isaaclab/isaaclab/envs/mdp/events.py [hunk 10]
```diff
@@ -1005,4 +1046,4 @@ def _randomize_prop_by_op(
         raise NotImplementedError(
             f"Unknown operation: '{operation}' for property randomization. Please use 'add', 'scale', or 'abs'."
         )
-    return data
+    return data
```
Comment (kellyguo11) [inline source/isaaclab/isaaclab/envs/mdp/events.py]: please run the formatter

Comment (shendredm) [reply source/isaaclab/isaaclab/envs/mdp/events.py]: Done.

File: source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/manager_based/locomotion/velocity/velocity_env_cfg.py (hunk 6 above)
Comment (shendredm) [reply source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/manager_based/locomotion/velocity/velocity_env_cfg.py]: I have added this here because this event runs at startup. I referred to several other IsaacGym RL papers and saw that this event is commonly used in training locomotion for legged robots.

File: source/isaaclab/isaaclab/envs/mdp/events.py [hunk 11]
```diff
@@ -376,11 +418,6 @@ def randomize_rigid_body_collider_offsets(
     """
     # extract the used quantities (to enable type-hinting)
     asset: RigidObject | Articulation = env.scene[asset_cfg.name]
-
```
Comment (Mayankm96) [inline source/isaaclab/isaaclab/envs/mdp/events.py]: Was this removed accidentally?

Comment (shendredm) [reply source/isaaclab/isaaclab/envs/mdp/events.py]: Yes, fixed it. Ready for Merge!!!

**Review by Mayankm96 — APPROVED** [review]

File: source/isaaclab/isaaclab/envs/mdp/events.py [hunk 12]
```diff
@@ -351,6 +351,49 @@ def randomize_rigid_body_mass(
...
+    # get the current com of the bodies (num_assets, num_bodies)
+    coms = asset.root_physx_view.get_coms().clone()
+
+    # Randomize the com in range
+    coms[:, body_ids, :3] += rand_samples
+
+    # Set the new coms
+    asset.root_physx_view.set_coms(coms, env_ids)
+
+
```
Comment (Mayankm96) [inline source/isaaclab/isaaclab/envs/mdp/events.py]:
```suggestion

```

File: source/isaaclab/isaaclab/envs/mdp/events.py [hunk 13]
```diff
@@ -351,6 +351,47 @@ def randomize_rigid_body_mass(
...
+    ).unsqueeze(1)
+
+    # get the current com of the bodies (num_assets, num_bodies)
+    coms = asset.root_physx_view.get_coms().clone()
+
+    # Randomize the com in range
+    coms[:, body_ids, :3] += rand_samples
+
+    # Set the new coms
+    asset.root_physx_view.set_coms(coms, env_ids)
```
Comment (Mayankm96) [inline source/isaaclab/isaaclab/envs/mdp/events.py:394]:
```suggestion
    asset.root_physx_view.set_coms(coms, env_ids)

```

File: source/isaaclab/isaaclab/envs/mdp/events.py [hunk 14]
```diff
@@ -351,6 +351,48 @@ def randomize_rigid_body_mass(
...
+
```
Comment (Mayankm96) [inline source/isaaclab/isaaclab/envs/mdp/events.py:395]:
```suggestion


```

**Comment by kellyguo11** [comment]
This change appears to have broken training for the locomotion environments. @shendredm could you double check the training? I'll revert the change in the environment for now.

**Comment by kellyguo11** [comment]
Awesome, thanks so much for the quick fix! I'll let CI run through the new PR and have it merged
